### PR TITLE
CPourTx structural changes

### LINKED ISSUE #799: select binary serialization in libsnark
Apparently it [defaults to] text-based. Ugh.

### LINKED ISSUE #927: Ensure ordering is correct for JoinSplit description fields
We currently serialize JoinSplit descriptions in the wrong order according to the spec. Once we're finished implementing all the fields in the spec, we should ensure the orderings are correct.

### PR DESCRIPTION
**Pull request by ebfull**
* Enable binary serialization of proofs and the proving key (closes #799) and make the proofs fixed-size.
* Reorder fields of CPourTx to match the spec (closes #927)

### CODE REVIEW

File: src/Makefile.am [hunk 1]
```diff
@@ -415,7 +415,7 @@ libzerocash_a_SOURCES = \
   zcash/prf.cpp \
   zcash/util.cpp
 
-libzerocash_a_CPPFLAGS = -fPIC -DCURVE_ALT_BN128 -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS $(HARDENED_CPPFLAGS) -std=c++11 -pipe -O2 -O0 -g -Wstack-protector -fstack-protector-all -fPIE -fvisibility=hidden -DSTATIC $(BITCOIN_INCLUDES)
+libzerocash_a_CPPFLAGS = -fPIC -DBINARY_OUTPUT -DCURVE_ALT_BN128 -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS $(HARDENED_CPPFLAGS) -pipe -O2 -O0 -g -Wstack-protector -fstack-protector-all -fPIE -fvisibility=hidden -DSTATIC $(BITCOIN_INCLUDES)
```
Comment (defuse) [inline src/Makefile.am]: Are we losing the `-std=c++11` flag here. It's set in the global `$(CXXFLAGS)` so does that mean it gets set for this build too?

File: src/test/sighash_tests.cpp [hunk 2]
```diff
@@ -135,14 +135,6 @@ void static RandomTransaction(CMutableTransaction &tx, bool fSingle) {
...
             pourtx.macs[0] = GetRandHash();
             pourtx.macs[1] = GetRandHash();
-            {
-                std::vector<unsigned char> txt;
-                int prooflen = insecure_rand() % 1000;
-                for (int i = 0; i < prooflen; i++) {
-                    txt.push_back(insecure_rand() % 256);
-                }
-                pourtx.proof = std::string(txt.begin(), txt.end());
-            }
```
Comment (defuse) [inline src/test/sighash_tests.cpp]: Should this be replaced by a call to `randombytes_buf()` filling the proof array?

File: src/zcash/JoinSplit.cpp [hunk 3]
```diff
@@ -266,8 +267,14 @@ class JoinSplitCircuit : public JoinSplit<NumInputs, NumOutputs> {
 
         std::stringstream ss;
         ss << proof;
+        std::string serialized_proof = ss.str();
 
-        return ss.str();
+        boost::array<unsigned char, ZKSNARK_PROOF_SIZE> result_proof;
+        //std::cout << "proof size in bytes when serialized: " << serialized_proof.size() << std::endl;
+        assert(serialized_proof.size() == ZKSNARK_PROOF_SIZE);
+        memcpy(&result_proof[0], &serialized_proof[0], ZKSNARK_PROOF_SIZE);
```
Comment (defuse) [inline src/zcash/JoinSplit.cpp]: I don't think this is correct: `serialized_proof` is a string, and strings don't have to store their contents contiguously in memory. I think you need to call `serialized_proof.c_str()` but I'm not sure how that deals with null characters.

Comment (defuse) [inline src/zcash/JoinSplit.cpp]: Never mind, std::string is a basic_string which says "The elements of a basic_string are stored contiguously, that is, for a basic_string s, &*(s.begin() + n) == &*s.begin() + n for any n in [0, s.size()), or, equivalently, a pointer to s[0] can be passed to functions that expect a pointer to the first element of a CharT[] array."  http://en.cppreference.com/w/cpp/string/basic_string

File: src/test/sighash_tests.cpp (hunk 2 above)
Comment (ebfull) [inline src/test/sighash_tests.cpp]: I'll add it.

File: src/zcash/JoinSplit.cpp (hunk 3 above)
Comment (defuse) [inline src/zcash/JoinSplit.cpp]: Note the above is only for C++11, before C++11 it was not true.

File: src/Makefile.am (hunk 1 above)
Comment (ebfull) [inline src/Makefile.am]: Yup!

**Comment by defuse** [comment]
ACK (but I left some comments).

**Comment by ebfull** [comment]
Rebased.

File: src/test/sighash_tests.cpp (hunk 2 above)
Comment (ebfull) [inline src/test/sighash_tests.cpp]: This is addressed!

**Comment by ebfull** [comment]
Rebased.

**Comment by defuse** [comment]
ACK!

**Comment by ebfull** [comment]
@zkbot r+

**Pull request merged**